central/declarativeconfig/README.md

```diff
@@ -0,0 +1,293 @@
+and storage types. 
+- This allows the UI to know whether an object is maintained by
+the declarative framework or by API users. It allows the UI to show the origin
+of displayed objects and to prevent users from trying to modify declaratively
+configured objects.
+The conversion functions live in the `central/convert` subdirectory.
+
+### Create data structures for YAML decoding
+These data structures are there to allow decoding the configuration files
+to process internal data structures.
+
+The protobuf enum types require dedicated decoders. Some of the existing types
+can be used as examples.
+
+The data structures for YAML decoding live in `pkg/declarativeconfig` and
+the top-level structure should implement the `Configuration` interface defined
+in `pkg/declarativeconfig/configuration.go`
+
+### Create data transformers to convert objects decoded from YAML to storage protobuf objects
+The transformer is in charge of converting the decoded YAML structure
+into one or more storage protobuf objects, sorted by type.
+The output of the transformer is then used to feed the updaters and populate
+the database.
+
+The transformers are defined in `pkg/declarativeconfig/transform` and should
+implement the `Transformer` interface defined in
+`pkg/declarativeconfig/transform/transformer.go`.
+
+### Add declarative constraints logic to the datastore layer
+This step is two-fold, impacts the `Upsert` method of the datastore
+and should fulfill two requirements:
+- Ensure the process that pushes an object for update is allowed to do so
+with regard to the origin of the upserted or deleted object.
+- Ensure the object references object of allowed origin
+(declaratively configured objects should not reference user-created objects).
+
+#### Ensure the process is allowed to push the requested object
+In the datastore `Upsert` method, apply the following code snipped to both
+the object passed as input and the object retrieved from the database (if any):
+```
+if !declarativeconfig.CanModifyResource(ctx, obj) {
+	return errox.NotAuthorized.CausedByf(
+		"object %q's origin is %s, cannot be modified or deleted with the current permission",
+		object.GetName(), object.GetTraits().GetOrigin()
+	)
+}
+```
+
+### Create declarative updaters
+The role of the updater is to upsert new or changed objects, as well as
+to remove objects which were removed from the declarative configuration files.
+The updater should implement the `ResourceUpdater` interface declared in
+`central/declarativeconfig/updater/updater.go`
+
+The `Upsert` method should be simple and straightforward, and delegate
+the action to the underlying datastore.
+
+The `DeleteResources` method should first list the items present
+in the database and candidate for removal, then attempt to remove them,
+and report removal errors in both its output and the declarative health
+datastore.
+
+### Register the above objects in the declarative framework
```

Comment:
Do we have a way to check/lint registration?

Reply:
ROX-30330